Added new kb article detect-row-detail-expansion-collapse-maui-datagrid (#873)
telerik/maui-docs: knowledge-base/detect-row-detail-expansion-collapse-maui-datagrid.md

---

> title: Detecting Row Detail Expansion or Collapse in DataGrid for MAUI

Reviewer: ```suggestion
title: Detecting If Row Detail Is Expanded or Collapsed in DataGrid
```

description: Learn how to identify when a row detail in the DataGrid for .NET MAUI is expanded or collapsed using MVVM.
type: how-to

> page_title: How to Detect Row Detail Expansion or Collapse in .NET MAUI DataGrid

Reviewer: ```suggestion
page_title: How to Detect Row Detail State in .NET MAUI DataGrid
```

slug: detect-row-detail-expansion-collapse-maui-datagrid
tags: datagrid, rowdetails, expandedrowdetails, toggle, mvvm, maui
res_type: kb
---

## Environment

| Product | Version | Author |
| --- | --- |
| DataGrid for .NET MAUI | 7.0.0 | [Dobrinka Yordanova](https://www.telerik.com/blogs/author/dobrinka-yordanova)| 

## Description

In a MAUI application using the DataGrid, I need to detect when a row detail is expanded or collapsed while following the MVVM pattern.

This KB article also answers the following questions:
- How can I listen to row detail state changes in a DataGrid?
- Is there a way to bind the expansion state of row details in DataGrid to a ViewModel?
- Can I use MVVM to track the expansion and collapse of row details in DataGrid?

## Solution

To detect when a row detail is expanded or collapsed in the [DataGrid]({%slug datagrid-overview%}) for MAUI using MVVM, you can employ the following approaches:

> * Working with the `ExpandedRowDetails` collection

Reviewer: ```suggestion
* Using the `ExpandedRowDetails` collection
```

    private void ExpandedItems_CollectionChanged(object? sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e)
    {
        if (e.Action == System.Collections.Specialized.NotifyCollectionChangedAction.Add)
        {
            // your logic here
        }
        else if(e.Action == System.Collections.Specialized.NotifyCollectionChangedAction.Remove)
        {
            // your logic here
        }
    }

    public ObservableCollection<Data> Items { get; set; }

    public ObservableCollection<Data> ExpandedItems { get; set; }
}

public class Data
{
    public string Country { get; set; }
    public string Capital { get; set; }
    public string Details { get; set; }
}
```

### Approach 2: Using the `ToggleRowDetailsButtonTap` Command

**1.** Utilize the `ToggleRowDetailsButtonTap` command provided by the DataGrid. This command is triggered whenever the toggle button for row details is tapped. In the command's execution logic, add a flag or a mechanism to check whether the item related to the toggled row detail is present in the `ExpandedRowDetails` collection. An item's presence indicates an expansion, while its absence indicates a collapse.

> Implement the necessary logic based on this condition to handle the expansion or collapse state.

Reviewer: Is this step 2? Or does it belong to step 1?

Reviewer 2: step 2

```C#
public class MyToggleRowDetailsButtonTapCommand : DataGridCommand
{
    public MyToggleRowDetailsButtonTapCommand()
    {
        Id = DataGridCommandId.ToggleRowDetailsButtonTap;
    }

    public override bool CanExecute(object parameter)
    {
        return true;
    }

    public override void Execute(object parameter)
    {
        // flag
        bool wasExpanded = this.Owner.ExpandedRowDetails.Contains(parameter);
        this.Owner.CommandService.ExecuteDefaultCommand(DataGridCommandId.ToggleRowDetailsButtonTap, parameter);
    }
}
```

> **3.** The `ViewModel` and data model definition:

Reviewer: Is this step 2 or 3?

Reviewer 2: step 2
